pennylane/math/single_dispatch.py

```diff
@@ -0,0 +1,327 @@
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Autoray registrations"""
+# pylint:disable=protected-access,import-outside-toplevel,wrong-import-position
+from importlib import import_module
+import numbers
+
+import autoray as ar
+import numpy as np
+
+
+def _i(name):
+    """Convenience function to import PennyLane
+    interfaces via a string pattern"""
+    if name == "tf":
+        return import_module("tensorflow")
+
+    if name == "qml":
+        return import_module("pennylane")
+
+    return import_module(name)
+
+
+# -------------------------------- NumPy --------------------------------- #
+from scipy.linalg import block_diag as _scipy_block_diag
+
+ar.register_function("numpy", "flatten", lambda x: x.flatten())
+ar.register_function("numpy", "coerce", lambda x: x)
+ar.register_function("numpy", "block_diag", lambda x: _scipy_block_diag(*x))
+ar.register_function("builtins", "block_diag", lambda x: _scipy_block_diag(*x))
+ar.register_function("numpy", "gather", lambda x, indices: x[np.array(indices)])
+
+
+def _scatter_element_add_numpy(tensor, index, value):
+    tensor[tuple(index)] += value
+    return tensor
+
+
+ar.register_function("numpy", "scatter_element_add", _scatter_element_add_numpy)
+
+
+# -------------------------------- Autograd --------------------------------- #
+
+
+# When autoray inspects PennyLane NumPy tensors, they will be associated with
+# the 'pennylane' module, and not autograd. Set an alias so it understands this is
+# simply autograd.
+ar.autoray._BACKEND_ALIASES["pennylane"] = "autograd"
+
+# When dispatching to autograd, ensure that autoray will instead call
+# qml.numpy rather than autograd.numpy, to take into account our autograd modification.
+ar.autoray._MODULE_ALIASES["autograd"] = "pennylane.numpy"
+
+
+ar.register_function("autograd", "flatten", lambda x: x.flatten())
+ar.register_function("autograd", "coerce", lambda x: x)
+ar.register_function("autograd", "block_diag", lambda x: _scipy_block_diag(*x))
+ar.register_function("autograd", "gather", lambda x, indices: x[np.array(indices)])
+
+
+def _to_numpy_autograd(x):
+    if hasattr(x, "_value"):
+        # Catches the edge case where the data is an Autograd arraybox,
+        # which only occurs during backpropagation.
+        return x._value
+
+    return x.numpy()
+
+
+ar.register_function("autograd", "to_numpy", _to_numpy_autograd)
+
+
+def _scatter_element_add_autograd(tensor, index, value):
+    size = tensor.size
+    flat_index = _i("qml").numpy.ravel_multi_index(index, tensor.shape)
+    t = [0] * size
+    t[flat_index] = value
+    return tensor + _i("qml").numpy.array(t).reshape(tensor.shape)

```

Comment:
Could we have a single call to `_i("qml")` and assign the result to an object at the beginning of the function, or do we make use of multiple calls to `import_module`?

Reply:
The nice thing about python imports is that the actual importation only happens the first time the module is imported. Further imports add no overhead :)

```python
import tensorflow as tf  # this will be slow
import tensorflow as tf  # instantaneous!
```